src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs

```diff
@@ -1169,6 +1169,8 @@ private int RunCore(TextWriter consoleOutput, ErrorLogger errorLogger, Cancellat
                             emitTestCoverageData: emitOptions.EmitTestCoverageData,
                             pePdbFilePath: emitOptions.PdbFilePath,
                             privateKeyOpt: privateKeyOpt,
+                            defaultSourceEncoding: Arguments.Encoding,
+                            fallbackSourceEncoding: compilation.SyntaxTrees.FirstOrDefault()?.Encoding,
```

Comment:
>compilation.SyntaxTrees.FirstOrDefault()?.Encoding [](start = 52, length = 50)

I don't think this is correct. Suppose that the first source file has a BOM and is UTF8 encoded, but the next file is SJIS encoded, which is the OS default. In that case we want to store SJIS as fallback encoding, not UTF8.

You'll need to update this code: [http://sourceroslyn.io/#Microsoft.CodeAnalysis/CommandLine/CommonCompiler.cs,265](http://sourceroslyn.io/#Microsoft.CodeAnalysis/CommandLine/CommonCompiler.cs,265)
I'd make [`EncodedStringText.Create`](http://sourceroslyn.io/#Microsoft.CodeAnalysis/EncodedStringText.cs,86) internal so that we can pass a custom `Lazy<Encoding>` to `getEncoding` parameter. Create and instance of `Lazy<Encoding>` in [CreateCompilation](http://sourceroslyn.io/#Microsoft.CodeAnalysis.CSharp/CommandLine/CSharpCompiler.cs,39) and pass it thru all the way to EncodedStringText.Create. After all trees are parsed you can check if the `Lazy<Encoding>` has value (has been initialized). If so then use that as the fallback encoding.

Reply:
on it, good catch